Tracking issue for `?` operator and `try` blocks (RFC 243, `question_mark` & `try_blocks` features)

Tracking issue for rust-lang/rfcs#243 and rust-lang/rfcs#1859.

See also the related tracking issue for `try_trait_v2`, https://github.com/rust-lang/rust/issues/84277 (https://github.com/rust-lang/rfcs/pull/3058)

For *heterogeneous* `try bikeshed MyType` blocks, see #149488

Implementation concerns:
- [x] `?` operator that is roughly equivalent to `try!` - #31954
- [x] `try { ... }` expression - https://github.com/rust-lang/rust/issues/39849
    - [x]  resolve `do catch { ... }` syntax question
        - Resolved as `try { .. }`, - https://github.com/rust-lang/rust/issues/50412
    - [x] resolve whether catch blocks should "wrap" result value (first addressed in https://github.com/rust-lang/rust/issues/41414, now being settled anew in https://github.com/rust-lang/rust/issues/70941)
    - [x] Add a test confirming that it's an `ExprWithBlock`, so works in a match arm without a comma
      - [x] https://github.com/rust-lang/rust/pull/120540
      - [x] https://github.com/rust-lang/rust/pull/148968
    - [ ] Address issues with type inference (`try { expr? }?` currently requires an explicit type annotation somewhere).
      - [x] https://github.com/rust-lang/rust/pull/148725
      - [ ] https://github.com/rust-lang/rfcs/pull/3721
- [x] settle design of the `Try` trait (https://github.com/rust-lang/rfcs/pull/1859)
    - [x] implement new `Try` trait (in place of `Carrier`) and convert `?` to use it (https://github.com/rust-lang/rust/pull/42275)
    - [x] add impls for `Option` and so forth, and a suitable family of tests (https://github.com/rust-lang/rust/pull/42526)
    - [x] improve error messages as described in the RFC (https://github.com/rust-lang/rust/issues/35946)
- [x] reserve `try` in new edition
- [x] block `try{}catch` (or other following idents) to leave design space open for the future, and point people to how to do what they want with `match` instead

I tried using this for the first time yesterday, and I was a little surprised that this:
```rust
#![feature(try_blocks)]
fn main() -> Result<(), ()> {
    let x: () = try {
        Err(())?
    }?;
    Ok(x)
}
```
does not compile due to 
```
error[E0284]: type annotations required: cannot resolve `<_ as std::ops::Try>::Ok == _`
``` 
Rather, I had to do 
```rust
#![feature(try_blocks)]
fn main() -> Result<(), ()> {
    let x: Result<(), ()> = try {
        Err(())?
    };
    let x = x?;
    Ok(x)
}
```
instead.

This was confusing at first, so maybe it is worth changing the error message or mentioning in `--explain`?

Has a syntax for type annotations been agreed? I was hoping for some `try { foo()?; bar()?; }.with_context(|_| failure::err_msg("foon' n' barn'")?;`, which is not even remotely interested in compiling: `error[E0282]: type annotations needed`.

https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=4e60d44a8f960cf03307a809e1a3b5f2

I wanted to give `try` a shot for some nested `Option` parsing as well:

```rust
#![feature(try_blocks)]

struct Config {
    log: Option<LogConfig>,
}

struct LogConfig {
    level: Option<String>,
}

fn example(config: &Config) {
    let x: &str = try { config.log?.level? }.unwrap_or("foo");
}
```

This fails with

```
error[E0282]: type annotations needed
  --> src/lib.rs:12:19
   |
12 |     let x: &str = try { config.log?.level? }.unwrap_or("foo");
   |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot infer type
   |
   = note: type must be known at this point
```

The closest I got was 

```rust
fn example(config: &Config) {
    let x: Option<&str> = try { &**config.log.as_ref()?.level.as_ref()? };
    let x = x.unwrap_or("foo");
}
```

The `as_ref` is quite unfortunate. I know that [`Option::deref`](https://github.com/rust-lang/rust/issues/50264) will help some here, but not enough. This feels like somehow match ergonomics )or related idea) should come into play.

The multiple lines is also unfortunate.

> a somewhat naive `try_` implementation

Yes, but your macro invocation returns a `String`, not a `&str`, requiring ownership. You don't show the surrounding code, but this will fail because we don't have ownership of the `Config`:

```rust
fn example(config: &Config) {
    let x = try_! { config.log?.level? }.unwrap_or_else(|| String::from("foo"));
}
```

```
error[E0507]: cannot move out of captured variable in an `Fn` closure
  --> src/lib.rs:20:21
   |
19 | fn example(config: &Config) {
   |            ------ captured outer variable
20 |     let x = try_! { config.log?.level? }.unwrap_or_else(|| String::from("foo"));
   |                     ^^^^^^^^^^ cannot move out of captured variable in an `Fn` closure
```

It also unconditionally allocates a `String`; I used `unwrap_or_else` in this example to avoid that inefficiency.

Apologies if this has been addressed already, but it seems strange to me that:

```rust
#![feature(try_blocks)]

fn main() -> Result<(), ()> {
    let result = try { // no type annotation
        Err(())?;
    };
    result.map_err(|err| err)
}
```
fails to compile with:
```
error[E0282]: type annotations needed
```
but:
```rust
#![feature(try_blocks)]

fn main() -> Result<(), ()> {
    let result : Result<_, _> = try { // partial type annotation
        Err(())?;
    };
    result.map_err(|err| err)
}
```
is okay.

If this were a problem because the type arguments of `Result` could not be deduced I'd understand, but, as shown above, that's not the case and rustc is able to perform inference once it is told that the result of a `try` expression is *some sort* of `Result`, which it ought to be able to infer from `core::ops::Try::into_result`.

Thoughts?

I thought I could use try blocks to easily add context to errors, but no luck so far.

I wrote a small function that works fine. Notice that `File::open()?` fails with a `std::io::Error` while the following line fails with an [`anyhow::Error`](https://docs.rs/anyhow/1.0.8/anyhow/struct.Error.html). Despite the differing types, the compiler figures out how to convert both to a `Result<_, anyhow::Error>`.

```rust
fn tls_add_cert(config: &ClientConfig, path: impl AsRef<Path>) -> Result<(usize, usize), anyhow::Error> {
    let path = path.as_ref();
    let mut file = BufReader::new(File::open(path)?);
    Ok(config.root_store.add_pem_file(&mut file)
        .map_err(|_| anyhow!("Bad PEM file"))?)
}
```

I wanted to add some error context so I tried to use a try-block and [anyhow's `with_context()`](https://docs.rs/anyhow/1.0.8/anyhow/trait.Context.html#tymethod.with_context):

```rust
fn tls_add_cert(config: &ClientConfig, path: impl AsRef<Path>) -> anyhow::Result<(usize, usize)> {
    let path = path.as_ref();
    try {
        let mut file = BufReader::new(File::open(path)?);
        Ok(config.root_store.add_pem_file(&mut file)
            .map_err(|_| anyhow!("Bad PEM file"))?)
    }
    .with_context(|| format!("Error adding certificate {}", path.display()))
}
```

But now type inference fails:

```rust
error[E0282]: type annotations needed
  --> src/net.rs:29:5
   |
29 | /     try {
30 | |         let mut file = BufReader::new(File::open(path)?);
31 | |         Ok(config.root_store.add_pem_file(&mut file)
32 | |             .map_err(|_| anyhow!("Bad PEM file"))?)
33 | |     }
   | |_____^ cannot infer type
   |
   = note: type must be known at this point
   ```

I don't understand why a type annotation is needed here but not in the first case. Nor do I see any easy way to add one, as opposed to using an [IIFE](https://en.wikipedia.org/wiki/Immediately_invoked_function_expression) which does let me add an annotation:

```rust
(|| -> Result<_, anyhow::Error> {
    let domain = DNSNameRef::try_from_ascii_str(host)?;
    let tcp = TcpStream::connect(&(host, port)).await?;

    Ok(tls.connect(domain, tcp).await?)
})()
.with_context(|| format!("Error connecting to {}:{}", host, port))
```

I don't know if this is a helpful example or just noise at this point, but I recently came across a case where the `try` block couldn't infer the type despite an explicit turbofish, and where the same code converted to an IIFE (which is the method I use on stable to get something akin to "try block behavior") works just fine. Top one gives compiler error, bottom one works:
```rust
fn using_try(file: File) {
    BufReader::new(file)
        .lines()
        .find_map(|line| try { Ok::<_, std::io::Error>(parse_line(line?)) }.transpose());
}

fn using_iife(file: File) {
    BufReader::new(file)
        .lines()
        .find_map(|line| (|| Ok::<_, std::io::Error>(parse_line(line?)))().transpose());
}
```
```
error[E0282]: type annotations needed
  --> src/lib.rs:10:26
   |
10 |         .find_map(|line| try { Ok::<_, std::io::Error>(parse_line(line?)) }.transpose());
   |                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot infer type
   |
   = note: type must be known at this point
```
https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=9b715a78fa6ea9017a1c2426789ae87d

I really don't know what more type annotation it wants, what I'm returning is unequivocally `Result<(), std::io::Error>`.

Not sure if it's a known issue but this [fails](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=ae55bf5f51edfde6e4cf2639aa2cc0b6) to compile on latest nightly compiler:


```rust
#![feature(try_blocks)]
fn foo(cond: bool) -> impl Iterator<Item=i32> {
    let result: Option<i32> = try {
        if cond {
            println!("Oops");
            return None;
        }
        55
    };
    result.into_iter()
}
```

```
error[E0277]: `Option<_>` is not an iterator
 --> src/lib.rs:2:23
  |
2 | fn foo(cond: bool) -> impl Iterator<Item=i32> {
  |                       ^^^^^^^^^^^^^^^^^^^^^^^ `Option<_>` is not an iterator
  |
  = help: the trait `Iterator` is not implemented for `Option<_>`
```

This might be an incorrect usage but at least error message might a clearer.

> error[E0277]: `Option<_>` is not an iterator

Doesn't look like an issue to me, the error message is correct: `Option` does not implement `Iterator`, only `IntoIterator`. And you can get the same error [without using `try_blocks`](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=1b50f007c3f8197552e8cd8ad75fa9c2).

BTW, the compiler goes on with:

```text
error[[E0308]](https://doc.rust-lang.org/nightly/error_codes/E0308.html): mismatched types
  --> src/lib.rs:10:5
   |
2  | fn foo(cond: bool) -> impl Iterator<Item=i32> {
   |                       ----------------------- expected because this return type...
...
6  |             return None;
   |                    ---- ...is found to be `Option<_>` here
...
10 |     result.into_iter()
   |     ^^^^^^^^^^^^^^^^^^ expected `Option<_>`, found `IntoIter<i32>`
   |
```

 So, you can easily fix this by adding `into_iter()` to your return statement: [playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=08390824b56e6467d60fe6c633cebb84)

Forgive me if this has been asked before. Are the details for the semantics of nested try blocks fully worked out? I have wrote a trivial example, but it should illustrate the question I am asking.

```rust
#![feature(try_blocks)]

use std::{fs::File, io::{self, BufRead, BufReader}, num::ParseIntError};

fn main() {
    let res: Result<u64, io::Error> = try {
        let file = File::options()
            .read(true)
            .open("number.txt")?;
        let mut buf = String::new();
        let res2: Result<u64, ParseIntError> = try {
            BufReader::new(file).read_line(&mut buf)?;
            buf.trim().parse()?
        };
        res2.expect("could not parse the number")
    };
    let num = res.expect("io error");
    println!("found num {num}");
}
```

This code does not compile, because the `read_line` function returns `Return<_, io::Error>`, but the inner try block expects a `Result<_, ParseIntError>`. This tells me that the current semantics are that the `?` operator will only look at the innermost try block.

While this example is a bit artificial, and I doubt whether nested try blocks should be considered, the same result can also be observed with a function that returns a `Result` which also contains a try block, as depicted in the following example.

```rust
#![feature(try_blocks)]

use std::{fs::File, io::{self, BufRead, BufReader}, num::ParseIntError};

fn parse_from_file() -> Result<u64, io::Error> {
    let file = File::options()
        .read(true)
        .open("number.txt")?;
    let mut buf = String::new();
    let res2: Result<u64, ParseIntError> = try {
        BufReader::new(file).read_line(&mut buf)?;
        buf.trim().parse()?
    };
    Ok(res2.expect("could not parse the number"))
}
```

This also fails to compile. Perhaps the semantics for nested try blocks and the like could be that it will check each try block from innermost to outermost for matches with the `?` operator, but this requires further discussion.

Finally, I want to point out an issue with regards to the diagnostic message of the compile error of the above function. I have included it below. In this example, the reason the code fails to compile is because it cannot convert the `io::Error` to `ParseIntError`. This occurs within the try block, so the diagnostic message should point to the line with the try block, `let res2: Result<u64, ParseIntError> = try {`. However, it instead points to the function declaration. I am not sure where this bug should be reported, so I included it here.

```
error[E0277]: `?` couldn't convert the error to `ParseIntError`
  --> main.rs:11:49
   |
5  | fn parse_from_file() -> Result<u64, io::Error> {
   |                         ---------------------- expected `ParseIntError` because of this
...
11 |         BufReader::new(file).read_line(&mut buf)?;
   |                              -------------------^ the trait `From<std::io::Error>` is not implemented for `ParseIntError`
   |                              |
   |                              this can't be annotated with `?` because it has type `Result<_, std::io::Error>`
   |
   = note: the question mark operation (`?`) implicitly performs a conversion on the error value using the `From` trait

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0277`.
```

--- a/compiler/rustc_feature/src/unstable.rs
+++ b/compiler/rustc_feature/src/unstable.rs
@@ -673,7 +673,7 @@ declare_features! (
     /// Allows inconsistent bounds in where clauses.
     (unstable, trivial_bounds, "1.28.0", Some(48214)),
     /// Allows using `try {...}` expressions.
-    (unstable, try_blocks, "1.29.0", Some(31436)),
+    (unstable, try_blocks, "1.29.0", Some(154391)),
     /// Allows using `try bikeshed TargetType {...}` expressions.
     (unstable, try_blocks_heterogeneous, "1.94.0", Some(149488)),
     /// Allows `impl Trait` to be used inside type aliases (RFC 2515).

--- a/src/doc/unstable-book/src/language-features/try-blocks.md
+++ b/src/doc/unstable-book/src/language-features/try-blocks.md
@@ -1,8 +1,8 @@
 # `try_blocks`
 
-The tracking issue for this feature is: [#31436]
+The tracking issue for this feature is: [#154391]
 
-[#31436]: https://github.com/rust-lang/rust/issues/31436
+[#154391]: https://github.com/rust-lang/rust/issues/154391
 
 ------------------------
 
@@ -14,14 +14,14 @@ block creates a new scope one can use the `?` operator in.
 
 use std::num::ParseIntError;
 
-let result: Result<i32, ParseIntError> = try {
+let result = try {
     "1".parse::<i32>()?
         + "2".parse::<i32>()?
         + "3".parse::<i32>()?
 };
 assert_eq!(result, Ok(6));
 
-let result: Result<i32, ParseIntError> = try {
+let result = try {
     "1".parse::<i32>()?
         + "foo".parse::<i32>()?
         + "3".parse::<i32>()?

--- a/src/tools/rust-analyzer/crates/ide-db/src/generated/lints.rs
+++ b/src/tools/rust-analyzer/crates/ide-db/src/generated/lints.rs
@@ -11576,9 +11576,9 @@ The tracking issue for this feature is: [#85731]
         label: "try_blocks",
         description: r##"# `try_blocks`
 
-The tracking issue for this feature is: [#31436]
+The tracking issue for this feature is: [#154391]
 
-[#31436]: https://github.com/rust-lang/rust/issues/31436
+[#154391]: https://github.com/rust-lang/rust/issues/154391
 
 ------------------------
 
@@ -11590,14 +11590,14 @@ block creates a new scope one can use the `?` operator in.
 
 use std::num::ParseIntError;
 
-let result: Result<i32, ParseIntError> = try {
+let result = try {
     "1".parse::<i32>()?
         + "2".parse::<i32>()?
         + "3".parse::<i32>()?
 };
 assert_eq!(result, Ok(6));
 
-let result: Result<i32, ParseIntError> = try {
+let result = try {
     "1".parse::<i32>()?
         + "foo".parse::<i32>()?
         + "3".parse::<i32>()?

--- a/tests/ui/README.md
+++ b/tests/ui/README.md
@@ -1462,7 +1462,7 @@ Tests for the `#[doc(hidden)]` items.
 
 ## `tests/ui/try-block/`
 
-`#![feature(try_blocks)]`. See [Tracking issue for `?` operator and `try` blocks (RFC 243, `question_mark` & `try_blocks` features)](https://github.com/rust-lang/rust/issues/31436).
+`#![feature(try_blocks)]` & `#![feature(try_blocks_heterogeneous)]`. See [Tracking Issue for homogeneous `try_blocks`](https://github.com/rust-lang/rust/issues/154391) & [Experimental Tracking Issue for Heterogeneous Try Blocks](https://github.com/rust-lang/rust/issues/149488).
 
 ## `tests/ui/try-trait/`
 

--- a/tests/ui/feature-gates/feature-gate-try_blocks.stderr
+++ b/tests/ui/feature-gates/feature-gate-try_blocks.stderr
@@ -8,7 +8,7 @@ LL | |         x
 LL | |     };
    | |_____^
    |
-   = note: see issue #31436 <https://github.com/rust-lang/rust/issues/31436> for more information
+   = note: see issue #154391 <https://github.com/rust-lang/rust/issues/154391> for more information
    = help: add `#![feature(try_blocks)]` to the crate attributes to enable
    = note: this compiler was built on YYYY-MM-DD; consider upgrading it if it is out of date
 

--- a/tests/ui/try-block/try-block-homogeneous-pre-expansion.stderr
+++ b/tests/ui/try-block/try-block-homogeneous-pre-expansion.stderr
@@ -4,7 +4,7 @@ warning: `try` blocks are unstable
 LL |     try {}
    |     ^^^^^^
    |
-   = note: see issue #31436 <https://github.com/rust-lang/rust/issues/31436> for more information
+   = note: see issue #154391 <https://github.com/rust-lang/rust/issues/154391> for more information
    = help: add `#![feature(try_blocks)]` to the crate attributes to enable
    = note: this compiler was built on YYYY-MM-DD; consider upgrading it if it is out of date
    = warning: unstable syntax can change at any point in the future, causing a hard error!

